retis/src/module/ct/bpf/ct.bpf.c

```diff
@@ -86,6 +87,34 @@ static __always_inline bool ct_protocol_is_supported(u16 l3num, u8 protonum)
 	return true;
 }
 
+/* See ctnetlink_dump_labels(). */
+static __always_inline void get_nf_ct_labels(struct ct_event *e,
+					     struct nf_conn *ct)
+{
+	struct nf_ct_ext *ext = BPF_CORE_READ(ct, ext);
+	struct nf_conn_labels *labels;
+	int offset, nf_ct_ext_labels;
+
+	/* Conntrack labels depend on CONFIG_NF_CONNTRACK_LABELS, the following
+	 * enum variant is only defined if enabled.
+	 */
+	if (!bpf_core_enum_value_exists(enum nf_ct_ext_id, NF_CT_EXT_LABELS))
+		return;
+
+	if (!ext)
+		return;
+
+	nf_ct_ext_labels = bpf_core_enum_value(enum nf_ct_ext_id, NF_CT_EXT_LABELS);
+	offset = BPF_CORE_READ(ext, offset)[nf_ct_ext_labels];
+	if (!offset)
+		return;
+
+	labels = (void *)ext + offset;
+
+	BUILD_BUG_ON(sizeof(labels->bits) != sizeof(e->labels));
```

Comment:
this should statically check at compile time against `vmlinux.h` definition.
```c
#define member_sz(type, member) sizeof(((type *)0)->member)
BUILD_BUG_ON(member_sz(struct nf_conn_labels, bits) != member_sz(struct ct_event, labels));
```
another approach can be the above putting `BUILD_BUG_ON()` below the `__binding` definition.
I slightly prefer having the static assertion close to the definition, but it's up to you.

That being said, I wonder if `bpf_core_field_size()` could be helpful here (although I think you're intentionally checking against `vmlinux.h`), so probably it's not the case.

Reply:
> this should statically check at compile time against `vmlinux.h` definition.

Isn't that what's already done here?

> ```c
> #define member_sz(type, member) sizeof(((type *)0)->member)
> BUILD_BUG_ON(member_sz(struct nf_conn_labels, bits) != member_sz(struct ct_event, labels));
> ```
> 
> another approach can be the above putting `BUILD_BUG_ON()` below the `__binding` definition. I slightly prefer having the static assertion close to the definition, but it's up to you.

I guess both make sense, close to the definition vs close to where the actual copy is made. The second option has the advantage of not requiring an additional (`member_sz`) definition.

> That being said, I wonder if `bpf_core_field_size()` could be helpful here (although I think you're intentionally checking against `vmlinux.h`), so probably it's not the case.

Yes, that's on purpose as there's no kernel version having different definitions so far. This is to prevent incompatibility when upgrading `vmlinux.h` in case this changes at some point. Catching runtime issues is not in the scope and many other places would be problematic as we can't check for every variable being read at runtime.

Reply:
> > this should statically check at compile time against `vmlinux.h` definition.
> 
> Isn't that what's already done here?
> 

Of course, it referred to your change. I rephrased the comment backward before posting and this probably was a bit unclear/redundant :)

> > ```c
> > #define member_sz(type, member) sizeof(((type *)0)->member)
> > BUILD_BUG_ON(member_sz(struct nf_conn_labels, bits) != member_sz(struct ct_event, labels));
> > ```
> >   
> > another approach can be the above putting `BUILD_BUG_ON()` below the `__binding` definition. I slightly prefer having the static assertion close to the definition, but it's up to you.
> 
> I guess both make sense, close to the definition vs close to where the actual copy is made. The second option has the advantage of not requiring an additional (`member_sz`) definition.
> 

There are of course advantages in the other way, but as I said, it's up to you. 

> > That being said, I wonder if `bpf_core_field_size()` could be helpful here (although I think you're intentionally checking against `vmlinux.h`), so probably it's not the case.
> 
> Yes, that's on purpose as there's no kernel version having different definitions so far. This is to prevent incompatibility when upgrading `vmlinux.h` in case this changes at some point. Catching runtime issues is not in the scope and many other places would be problematic as we can't check for every variable being read at runtime.

many other places would have been my follow-up :)
Just checking.

Reply:
> > > this should statically check at compile time against `vmlinux.h` definition.
> > 
> > Isn't that what's already done here?
> 
> Of course, it referred to your change. I rephrased the comment backward before posting and this probably was a bit unclear/redundant :)

Thanks for clarifying, I was not sure if you meant this or if I was missing something :)

> > > ```c
> > > #define member_sz(type, member) sizeof(((type *)0)->member)
> > > BUILD_BUG_ON(member_sz(struct nf_conn_labels, bits) != member_sz(struct ct_event, labels));
> > > ```
> > >   
> > > another approach can be the above putting `BUILD_BUG_ON()` below the `__binding` definition. I slightly prefer having the static assertion close to the definition, but it's up to you.
> > 
> > 
> > I guess both make sense, close to the definition vs close to where the actual copy is made. The second option has the advantage of not requiring an additional (`member_sz`) definition.
> 
> There are of course advantages in the other way, but as I said, it's up to you.

Let's not introduce another helper then, but overall I don't have a strong preference.

> > > That being said, I wonder if `bpf_core_field_size()` could be helpful here (although I think you're intentionally checking against `vmlinux.h`), so probably it's not the case.
> > 
> > Yes, that's on purpose as there's no kernel version having different definitions so far. This is to prevent incompatibility when upgrading `vmlinux.h` in case this changes at some point. Catching runtime issues is not in the scope and many other places would be problematic as we can't check for every variable being read at runtime.
> 
> many other places would have been my follow-up :) Just checking.

Every bit copied from one place to an event can be checked, although not all copies can be problematic. This specific case could benefit from a runtime check, but I guess it would just flag one of the many possible incompatibilities we could see with newer kernels. In such cases all those things are just bugs we'll have to fix anyway. No strong opinion on this too.

I did check other places when doing this patch and the other ones where:
- IP src/dst
- Ports
- ICMP id

All those are protocol-related, I don't think a check is needed.

Reply:
yep, I agree (especially with the "every place read at runtime" part).